presto-main/src/main/java/io/prestosql/eventlistener/EventListenerManager.java

```diff
@@ -55,59 +56,76 @@ public void addEventListenerFactory(EventListenerFactory eventListenerFactory)
+                Map<String, String> instanceProperties = ImmutableMap.copyOf(loadProperties(instanceConfigFile));
+                addConfiguredEventListener(instanceName, instanceProperties);
+            }
+        } else {
+            checkArgument(!isNullOrEmpty(eventListenerName),
+                "EventListener plugin configuration for %s does not contain %s", EVENT_LISTENER_CONFIGURATION_FILE.getAbsoluteFile(), EVENT_LISTENER_PROPERTY_NAME);
+            checkArgument(!isNullOrEmpty(eventListenerInstanceNames),
+                "Event listener configuration %s does not contain %s", EVENT_LISTENER_CONFIGURATION_FILE.getAbsoluteFile(), EVENT_LISTENER_INSTANCE_NAMES);
+        }
     }
 
     @VisibleForTesting
-    protected void setConfiguredEventListener(String name, Map<String, String> properties)
+    protected void addConfiguredEventListener(String instanceName, Map<String, String> properties)
     {
-        requireNonNull(name, "name is null");
+        requireNonNull(instanceName, "instanceName is null");
         requireNonNull(properties, "properties is null");
 
+        String eventListenerName = properties.get(EVENT_LISTENER_PROPERTY_NAME);
+        checkArgument(!isNullOrEmpty(eventListenerName),
+            "EventListener plugin configuration for %s does not contain %s", instanceName, EVENT_LISTENER_PROPERTY_NAME);
+
         log.info("-- Loading event listener --");
 
-        EventListenerFactory eventListenerFactory = eventListenerFactories.get(name);
-        checkState(eventListenerFactory != null, "Event listener '%s' is not registered", name);
+        checkState(eventListenerFactories.containsKey(eventListenerName), "Event listener factory for %s is not registered", eventListenerName);
+        EventListenerFactory eventListenerFactory = eventListenerFactories.get(eventListenerName);
 
-        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(eventListenerFactory.getClass().getClassLoader())) {
-            EventListener eventListener = eventListenerFactory.create(ImmutableMap.copyOf(properties));
-            this.configuredEventListener.set(Optional.of(eventListener));
-        }
+        EventListener eventListener = eventListenerFactory.create(properties);
+        Map<String, EventListener> eventListeners = configuredEventListeners.get();
+        eventListeners.put(instanceName, eventListener);
 
-        log.info("-- Loaded event listener %s --", name);
+        configuredEventListeners.compareAndSet(configuredEventListeners.get(), eventListeners);
+        log.info("-- Loaded event listener %s --", instanceName);
     }
 
     public void queryCompleted(QueryCompletedEvent queryCompletedEvent)
     {
-        if (configuredEventListener.get().isPresent()) {
-            configuredEventListener.get().get().queryCompleted(queryCompletedEvent);
+        for (EventListener listener : configuredEventListeners.get().values()) {
+            listener.queryCompleted(queryCompletedEvent);
```

Comment:
I’m thinking we should catch exceptions and log a warning, so that one listener doesn’t affect others (especially since there is no defined order they will be invoked)

Reply:
Added a warning for when exceptions are thrown from the event listeners but I believe this would still allow event listeners to block one another

Reply:
Yes, being slow or hanging is a problem, but there's not much that can be done. We'd need separate executors for each listener with a fixed queue size that eventually discards events.